pkg/controller/nodeobservatory/controller.go

```diff
@@ -0,0 +1,379 @@
+package nodeobservatory
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	api_v1 "k8s.io/client-go/pkg/api/v1"
+
+	"github.com/sapcc/kubernikus/pkg/apis/kubernikus/v1"
+	"github.com/sapcc/kubernikus/pkg/controller/config"
+)
+
+const (
+	BaseDelay              = 5 * time.Second
+	MaxDelay               = 300 * time.Second
+	KlusterRecheckInterval = 5 * time.Minute
+)
+
+type (
+	AddFunc    func(kluster *v1.Kluster, node *api_v1.Node)
+	UpdateFunc func(kluster *v1.Kluster, nodeCur, nodeOld *api_v1.Node)
+	DeleteFunc func(kluster *v1.Kluster, node *api_v1.Node)
+
+	NodeObservatory struct {
+		Controller
+
+		config.Factories
+		config.Clients
+		queue               workqueue.RateLimitingInterface
+		logger              log.Logger
+		nodeInformerMap     sync.Map
+		handlersMux         sync.RWMutex
+		addEventHandlers    []AddFunc
+		updateEventHandlers []UpdateFunc
+		deleteEventHandlers []DeleteFunc
+		stopCh              <-chan struct{}
+		informerWg          *sync.WaitGroup
+	}
+)
+
+func NewController(factories config.Factories, clients config.Clients, logger log.Logger) *NodeObservatory {
+	logger = log.With(logger,
+		"controller", "nodeobservatory",
+	)
+
+	controller := &NodeObservatory{
+		Factories:       factories,
+		Clients:         clients,
+		queue:           workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(BaseDelay, MaxDelay)),
+		logger:          logger,
+		nodeInformerMap: sync.Map{},
+	}
+
+	controller.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.klusterAddFunc,
+		UpdateFunc: controller.klusterUpdateFunc,
+		DeleteFunc: controller.klusterDeleteFunc,
+	})
+
+	return controller
+}
+
+func (n *NodeObservatory) Run(threadiness int, stopCh <-chan struct{}, wg *sync.WaitGroup) {
+	n.logger.Log(
+		"msg", "starting run loop",
+		"threadiness", threadiness,
+		"v", 2,
+	)
+
+	n.stopCh = stopCh
+	n.informerWg = wg
+
+	defer n.queue.ShutDown()
+	defer wg.Done()
+	wg.Add(1)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stopCh)
+	}
+
+	ticker := time.NewTicker(KlusterRecheckInterval)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				n.requeueAllKlusters()
+			case <-stopCh:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+	<-stopCh
+}
+
+func (n *NodeObservatory) requeueAllKlusters() (err error) {
+	defer func() {
+		n.logger.Log(
+			"msg", "requeued all",
+			"v", 1,
+			"err", err,
+		)
+	}()
+
+	klusters, err := n.Factories.Kubernikus.Kubernikus().V1().Klusters().Lister().List(labels.Everything())
```

Comment:
Is this lister confined to the namespace used by the operator?

Reply:
I think klusters from all ns are listed.

Reply:
Ok that is not correct and needs to be changed. The `NodeObservatory` should only consider Klusters in the given namespace.

Reply:
Scratch that. The factory is confined to the namespace used by the operator. [See here]( https://github.com/sapcc/kubernikus/blob/master/pkg/controller/operator.go#L223)